src/std_extensions/collections.rs

```diff
@@ -31,13 +31,13 @@ pub const EXTENSION_NAME: ExtensionId = ExtensionId::new_unchecked("Collections"
 
 #[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
 /// Dynamically sized list of values, all of the same type.
-pub struct ListValue(Vec<Value>);
+pub struct ListValue(Vec<Value>, Type);
 
 impl ListValue {
     /// Create a new [CustomConst] for a list of values.
     /// (The caller will need these to all be of the same type, but that is not checked here.)
-    pub fn new(contents: Vec<Value>) -> Self {
-        Self(contents)
+    pub fn new(t: Type, contents: Vec<Value>) -> Self {
```

Comment:
I would move in favour of two constructors:
* `new_empty` that takes just a Type
* `new` that takes a Vec<Value> and not a type, and extracts the type from the values (checking that they are all the same, I guess - I think it's reasonable to panic if they are not)

Reply:
We do not seem to have an interface to get the type of a `Value`, although we can do it now. I will add.

Reply:
Hmm, not so easy. a `Value::Function` can be done via `hugr.get_function_signature().unwrap()`, although I'm not sure this is a good idea.
`Value::Sum` does not know the full type.

I'm going to leave this for now, so your second bullet will not be satisfied. i.e. no checking they are all the same. Lmk if you disagree.